lib/multiplexer/multiplexer.go

```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
+					return nil, trace.BadParameter(duplicateProxyLineError)
+				}
 			}
+
+			proxyLine = newPROXYLine
+
 			// repeat the cycle to detect the protocol
 		case ProtoProxyV2:
-			newProxyLine, err := ReadProxyLineV2(reader)
+			newPROXYLine, err := ReadProxyLineV2(reader)
 			if err != nil {
 				return nil, trace.Wrap(err, invalidProxyV2LineError)
 			}
-			if newProxyLine == nil {
-				continue
+			if newPROXYLine == nil {
+				unsignedPROXYLineReceived = true
+				continue // Skipping LOCAL command of PROXY protocol
 			}
 
 			// If proxyline is not signed, so we don't try to verify to avoid unnecessary load
-			if m.CertAuthorityGetter != nil && m.LocalClusterName != "" && newProxyLine.IsSigned() {
-				err = newProxyLine.VerifySignature(m.context, m.CertAuthorityGetter, m.LocalClusterName, m.Clock)
+			if m.CertAuthorityGetter != nil && m.LocalClusterName != "" && newPROXYLine.IsSigned() {
+				err = newPROXYLine.VerifySignature(m.context, m.CertAuthorityGetter, m.LocalClusterName, m.Clock)
 				if errors.Is(err, ErrNoHostCA) {
 					m.WithFields(log.Fields{
 						"src_addr": conn.RemoteAddr(),
 						"dst_addr": conn.LocalAddr(),
 					}).Warnf("%s - could not get host CA", invalidProxySignatureError)
 					continue
 				}
-				// DELETE IN 14.0, early 12 versions could send PROXY headers to remote auth server
-				if errors.Is(err, ErrNonLocalCluster) {
-					m.WithFields(log.Fields{
-						"src_addr": conn.RemoteAddr(),
-						"dst_addr": conn.LocalAddr(),
-					}).Debugf("%s - signed by non local cluster", invalidProxySignatureError)
-					continue
-				}
 				if err != nil {
 					return nil, trace.Wrap(err, "%s %s -> %s", invalidProxySignatureError, conn.RemoteAddr(), conn.LocalAddr())
 				}
 				m.WithFields(log.Fields{
 					"conn_src_addr":   conn.RemoteAddr(),
 					"conn_dst_addr":   conn.LocalAddr(),
-					"client_src_addr": newProxyLine.Source.String(),
+					"client_src_addr": newPROXYLine.Source.String(),
 				}).Tracef("Successfully verified signed PROXYv2 header")
 			}
 
 			// If proxy line is signed and successfully verified and there's no already signed proxy header,
 			// we accept, otherwise reject
-			if newProxyLine.IsVerified {
+			if newPROXYLine.IsVerified {
 				if proxyLine != nil && proxyLine.IsVerified {
 					return nil, trace.BadParameter(duplicateSignedProxyLineError)
 				}
 
-				proxyLine = newProxyLine
+				proxyLine = newPROXYLine
 				continue
 			}
 
-			if m.CertAuthorityGetter != nil && newProxyLine.IsSigned() && !newProxyLine.IsVerified {
-				return nil, trace.BadParameter("could not verify proxy line signature")
+			if m.CertAuthorityGetter != nil && newPROXYLine.IsSigned() && !newPROXYLine.IsVerified {
+				return nil, trace.BadParameter("could not verify PROXY line signature")
 			}
 
 			// This is unsigned proxy line, return error if external PROXY protocol is not enabled
-			if !m.EnableExternalProxyProtocol {
+			if m.PROXYProtocolMode == PROXYProtocolOff {
```

Comment:
Why isn't this done earlier? Are signed PROXYv2 headers always allowed, even if `proxy_protocol` is set to `off`? That wasn't clear to me from the RFC update.

Reply:
Yep, signed PROXY headers are always allowed and here, as comment says we enter territory of unsigned header. RFD mentions in the scope that only unsigned PROXY headers are the focus of the RFD.